crates/event-queue/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+//! # Event Queue
+//!
+//! This crate provides a thread-safe event queue system for collecting and managing various types
+//! of events that occur during GraphQL gateway operation execution.
+//!
+//! The event queue stores events only if the gateway is started with a hooks extension, which defines
+//! an event filter in its configuration. Otherwise every event push is a no-op.
+
+mod events;
+
+pub use events::*;
+
+use crossbeam_queue::SegQueue;
+use extension_catalog::{EventFilter, EventFilterType, Extension};
+
+/// A thread-safe queue for collecting events during gateway operation execution.
+///
+/// The `EventQueue` provides a centralized mechanism for collecting various types of events
+/// that occur during request processing. It uses a lock-free queue internally for efficient
+/// concurrent access.
+///
+/// Events are only collected if an extension is configured and interested in receiving them,
+/// as determined by the event filter configuration.
+pub struct EventQueue {
+    inner: Option<SegQueue<Event>>,
+    filter: Option<EventFilter>,
+}
+
+impl EventQueue {
+    /// Creates a new `EventQueue` instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `hooks` - Optional extension configuration that determines whether events should be
+    ///   collected and which types of events to filter.
+    ///
+    /// # Returns
+    ///
+    /// A new `EventQueue` instance. If no extension is provided, the queue will be inactive
+    /// and won't collect any events.
+    pub fn new(hooks: &Option<Extension>) -> Self {
+        let (inner, filter) = match hooks {
+            Some(hooks) => {
+                let filter = hooks.manifest.event_filter;
+                (Some(SegQueue::new()), filter)
+            }
+            None => (None, None),
+        };
+
+        Self { inner, filter }
+    }
+
+    /// Pushes a GraphQL operation execution event to the queue.
+    ///
+    /// This method queues an event representing the execution of a GraphQL operation,
+    /// including details like the operation name, document, execution duration, and response status.
+    ///
+    /// The event is only queued if operation events are allowed by the current filter configuration.
+    /// The builder is only consumed and built if the event will actually be pushed, avoiding
+    /// expensive operations when events are filtered out.
+    ///
+    /// # Arguments
+    ///
+    /// * `builder` - The operation execution event builder
+    pub fn push_operation(&self, builder: ExecutedOperationBuilder<'_>) {
+        if !self.event_allowed(EventFilterType::Operation) {
+            return;
+        }
+
+        let Some(queue) = self.inner.as_ref() else {
```

Comment:
This condition seems a bit unnecessary now as the `filter` will be `None`, it's probably cheap enough to just create a queue each time. Or `filter` being `None` is redundant.